### DISCUSSION TITLE
PHP-FPM high/100% CPU usage after deploying

### DISCUSSION THREAD

**Discussion by chrisloftus** [deployphp/deployer#3907] [Bugs]
### Deployer Version

v7.4.0

### Target OS

Ubuntu 22.04

### Which PHP version are you using?

PHP 8.3

### Content of deploy.php or deploy.yaml

```php
<?php

namespace Deployer;

require 'recipe/laravel.php';

set('bin/php', function () {
    return '/usr/bin/php8.3';
});

// Project name
set('application', 'example.com');

set('branch', 'master');

// Project repository
set('repository', 'git@github.com:example/example.git');

// [Optional] Allocate tty for git clone. Default value is false.
set('git_tty', true);

// Shared files/dirs between deploys
// add('shared_files', ['public/sitemap.xml']);
add('shared_dirs', []);

// Writable dirs by web server
add('writable_dirs', []);

// Hosts

// host('receiptagent')
//     ->set('deploy_path', '~/{{application}}');

host('example.com')
    ->set('deploy_path', '~/web')
    ->setRemoteUser('admin')
    ->set('bin/php', fn () => '/usr/bin/php');

// Tasks

task('reload:php-fpm', function () {
    run('sudo systemctl reload php8.3-fpm.service');
})->select('atp');

// [Optional] if deploy fails automatically unlock.
after('deploy:failed', 'deploy:unlock');
```

### Steps to reproduce

I used to reload FPM after a deploy but did not like this since it can kill requests.

I changed nginx to use `$realpath_root` to avoid having to reload FPM.

Since changing this, every now and then a deploy will trigger a spike in FPM CPU usage and go to 80% - 100% CPU usage for long periods.

Reloading FPM fixes this.

I suspect the FPM reload was masking this issue when I had it in the deploy script.

Does anyone know why this might happen?

All I see in the FPM logs is the increase pm.max_children setting, but this happens at other times and monitoring the app / server in New Relic is barely using any resources. So I'm not worried about that.

**Comment by antonmedv** [comment]
You should check the following configurations:

1. **Nginx Configuration:**
   - Ensure the `$realpath_root` directive is correctly configured.
   - Verify that the `fastcgi_pass` and related directives are properly set.

2. **PHP-FPM Configuration:**
   - Check `pm.max_children` and other `pm.*` settings to ensure they are appropriate for your workload.
   - Review `php-fpm.conf` and pool configuration files for any anomalies.

3. **Deployment Configuration:**
   - Ensure that the deploy script and any post-deploy tasks are properly configured and do not interfere with PHP-FPM operations.

**Comment by chrisloftus** [comment]
I've checked nginx configuration and increased pm.* settings and now do not see any warnings in FPM logs at all.

However when I deploy the CPU usage is still spiking occasionally.

Deployed at 11:45. Response time increased 4x post deploy. Manually reloaded FPM @ 11:55.

<img width="718" alt="Screenshot 2024-10-05 at 1 03 06 pm" src="https://github.com/user-attachments/assets/aa16f55a-3eeb-4f78-a769-c83dcc47b942">

Appreciate this might not be a deployer issue but it is strange how it only happens after deploy sometimes.
